ocaml/xenopsd/xc/device.ml

```diff
@@ -92,6 +90,170 @@ module Profile = struct
         Qemu_upstream_compat
 end
 
+(** File-descriptor event monitor implementation for the epoll library *)
+module Monitor = struct
+  let create () = Polly.create ()
+
+  let add m fd = Polly.add m fd Polly.Events.inp
+
+  let remove m fd = Polly.del m fd
+end
+
+(* Helper module to start up services *)
+module Service = struct
+  type t = {
+      name: string
+    ; domid: domid
+    ; exec_path: string
+    ; pid_path: string
+    ; cancel_path: string
+    ; timeout_seconds: float
+    ; args: string list
+    ; execute: path:string -> args:string list -> domid:domid -> unit -> string
+  }
+
+  let alive service =
+    let is_active = Fe_systemctl.is_active ~service in
+    ( if not is_active then
+        let status = Fe_systemctl.show ~service in
+        error
+          "%s: unexpected termination \
+           (Result=%s,ExecMainPID=%d,ExecMainStatus=%d,ActiveState=%s)"
+          service status.result status.exec_main_pid status.exec_main_status
+          status.active_state
+    ) ;
+    is_active
+
+  type watch_trigger = Finished | Cancelled | Empty
+
+  exception ECancelled of Xenops_task.task_handle
+
+  let raise_e = function
+    | ECancelled t ->
```

Comment:
Typo here? What is `ECancelled` vs `Cancelled`?

Reply:
`ECancelled` is correct, we're matching exceptions. I would have like to not have to define `ECancelled` not `raise_e`, but `xcp-idl`'s `Xenops_task` is obtuse and makes it difficult to actually expose the type of exception, so `raise_cancelled` is used instead